src/Token/TokenProvider.php

```diff
@@ -0,0 +1,279 @@
+<?php
+namespace Aws\Token;
+
+use Aws;
+use Aws\Api\DateTimeResult;
+use Aws\CacheInterface;
+use Aws\Exception\TokenException;
+use GuzzleHttp\Promise;
+
+/**
+ * Token providers are functions that accept no arguments and return a
+ * promise that is fulfilled with an {@see \Aws\Token\TokenInterface}
+ * or rejected with an {@see \Aws\Exception\TokenException}.
+ *
+ * <code>
+ * use Aws\Token\TokenProvider;
+ * $provider = TokenProvider::defaultProvider();
+ * // Returns a TokenInterface or throws.
+ * $token = $provider()->wait();
+ * </code>
+ *
+ * Token providers can be composed to create a token using conditional
+ * logic that can create different tokens in different environments. You
+ * can compose multiple providers into a single provider using
+ * {@see Aws\Token\TokenProvider::chain}. This function accepts
+ * providers as variadic arguments and returns a new function that will invoke
+ * each provider until a token is successfully returned.
+ */
+class TokenProvider
+{
+    use ParsesIniTrait;
+    const ENV_PROFILE = 'AWS_PROFILE';
+
+    /**
+     * Create a default token provider tha checks for cached a SSO token from
+     * the CLI
+     *
+     * This provider is automatically wrapped in a memoize function that caches
+     * previously provided tokens.
+     *
+     * @param array $config Optional array of token provider options.
+     *
+     * @return callable
+     */
+    public static function defaultProvider(array $config = [])
+    {
+
+        $cacheable = [
+            'sso',
+        ];
+
+        $defaultChain = [
```

Comment:
Mainly asking because I'm unfamiliar with the spec— is this chain in place for extensibility? i.e. other provider types being added in the future?